nixos/modules/programs/openvpn3.nix

```diff
@@ -1,43 +1,42 @@
 { config, lib, pkgs, ... }:
 
-let
-  cfg = config.programs.openvpn3;
-in
-{
+let cfg = config.programs.openvpn3;
+in {
   options.programs.openvpn3 = {
     enable = lib.mkEnableOption "the openvpn3 client";
     package = lib.mkOption {
       type = lib.types.package;
       default = pkgs.openvpn3.override {
         enableSystemdResolved = config.services.resolved.enable;
       };
-      defaultText = lib.literalExpression ''pkgs.openvpn3.override {
-        enableSystemdResolved = config.services.resolved.enable;
-      }'';
+      defaultText = lib.literalExpression ''
+        pkgs.openvpn3.override {
+          enableSystemdResolved = config.services.resolved.enable;
+        }
+      '';
       description = ''
         Which package to use for `openvpn3`.
       '';
     };
   };
 
   config = lib.mkIf cfg.enable {
-    services.dbus.packages = [
-      cfg.package
-    ];
+    services.dbus.packages = [ cfg.package ];
 
     users.users.openvpn = {
       isSystemUser = true;
       uid = config.ids.uids.openvpn;
       group = "openvpn";
     };
 
-    users.groups.openvpn = {
-      gid = config.ids.gids.openvpn;
-    };
+    users.groups.openvpn = { gid = config.ids.gids.openvpn; };
 
-    environment.systemPackages = [
-      cfg.package
-    ];
-  };
+    environment.systemPackages = [ cfg.package ];
+
+    systemd.packages = [ cfg.package ];
 
+    system.activationScripts.openvpn3-admin-init-config = ''
```

Comment:
Shouldn't it be moved to ExecStartPre?
Otherwise, this activation script requires dependencies, as some setups might not allow writing to some directories on this point (E.g impermanence)
Or maybe user should call config initialization by themselves?

Reply:
> Shouldn't it be moved to ExecStartPre?

Sounds right, I will try doing this.

> Or maybe user should call config initialization by themselves?

I don't think so: without this step `openvpn3` will not work if, for instance, `systemd-resolved` is enabled, but I would expect it to observe the change in the configuration.

Reply:
@dsommers, could you please explain what is the right lifecycle of the config file? Is it only modified by `openvpn3-admin init-config --write-config` or there are other legitimate ways to change its content?
Should we provide some customization to it, or we should just rely on its detection mechanisms?

Reply:
There are currently 2 config files in `/var/lib/openvpn3`.  The `netcfg.json` and `log-service.json` files; which are used by the `openvpn3-service-netcfg` and `openvpn3-service-log` services. But it may come more config files in later releases.

These config files are normally to be modified by `openvpn3-admin` via the `log-service` and `netcfg-service` "commands".

The `openvpn3-admin init-config` will by default not modify any configuration files, unless `--write-configs` is provided.  But it will never overwrite any existing configuration settings unless `--force ` is provided.  It is therefore safe to run `openvpn3-admin init-config --write-configs` during the install phase.  For fresh installs, the system will be probed for what is considered to be the best settings for OpenVPN 3 Linux.  For version updates, the existing configs will normally not change existing settings.

In the future, the `openvpn3-admin init-config --write-configs` may upgrade existing configuration files if needed (if settings changes names, values or "types"), but the main goal will be to try to preserve the existing behaviour as far as possible.  To overwrite existing settings with a potentially different behaviour should require `--force`.

Reply:
In regards to moving it to `ExecStartPre` .... OpenVPN 3 Linux does not ship any systemd unit files for the main services.  The `openvpn3-admin` is a stand-alone binary used for management purposes only.  In some cases it doesn't even depend on the OpenVPN 3 Linux services running (like the `openvpn3-admin journal` command).

The only unit file which is provided is `openvpn3-session@.service`.  That unit is just a wrapper around the `openvpn3-systemd` script to start VPN sessions via systemd (typically during boot).  This unit and the `openvpn3-systemd` script should not have any ties to the OpenVPN 3 Linux services other than being a consumer of the services provided by that stack.  It should definitely not configure the OpenVPN 3 Linux services.

Reply:
Thanks a lot for your detailed answer!

In this case, I would currently stick to the approach in which the config is re-created (`--force`) on `nixos-rebuild` since it will allow it to pick changes in things like `resolved` (it can be changed via NixOS config). If any user-configurable options will appear in the future, it will probably be possible to expose them via module configuration.

In this case, would you please take a look at the corresponding systemd service in `nixos/modules/programs/openvpn3.nix` (permalink for the current commit):

https://github.com/NixOS/nixpkgs/blob/08d153fb197741c90f93be53510d6f863f9e2f94/nixos/modules/programs/openvpn3.nix#L40-L54

It does not currently work <img src="https://cdn2.tgramsearch.com/images/3/2/329cb400773f4a0f0c0e64fa29bda178.jpg" alt="Harold" width="20px"> but I am working on it, so if there are any advice on it then it would be appreciated.

AFAIKS, being a soft dependency of `openvpn3-session@.service` is not right (it does not get activated when needed) so I am thinking on a better trigger for it.

Reply:
I have no real understanding of how NixOS does things, so this looks like some black magic to me.

Reply:
Fair enough)

It desugars into the following systemd `system/` service:

```systemd
[Unit]
After=systemd-resolved.service
Before=openvpn3-session@.service
Description=Generates openvpn3 configuration files

[Service]
# A few autp-generated [Environment]
ExecStart=/nix/store/.../bin/openvpn3-admin init-config --write-configs --force
RemainAfterExit=yes
Type=oneshot

[Install]
WantedBy=openvpn3-session@.service
```

Reply:
That will only work if the VPN sessions are started using the `openvpn3-sessions@.service` unit.  There are no guarantee VPN sessions will only be started via this systemd unit.

VPN sessions can be started using the `openvpn3 session-start` command, or via third-party tools like [openvpn3-indicator](https://github.com/OpenVPN/openvpn3-indicator/), not being managed by systemd at all.  In our roadmap we also have NetworkManager integration too.

That's why package install/update is the best place to do this call; not at runtime with a VPN session start.

Reply:
> That's why package install/update is the best place to do this call; not at runtime with a VPN session start.

Unfortunately, there is no package installation step in nixos, however the whole OS configuration is declarative.

Probably it would be best the to directly generate those config files based on system configuration values?
journald is kinda mandatory in nixos, and only remaining (for now) config value is resolvd support, which users won't be toggling frequently. If we store those configs in `/etc`, then we get to declaratively define configuration inside (Using `environment.etc."openvpn3/netconf.json"` option).
One caveat is that it might be possible that `openvpn3` package gets updated, and then some users might experience crash due to outdated config files, but it is possible to write NixOS test to catch those issues.

Reply:
OpenVPN 3 Linux used some "expected distro defaults" detected at compile time earlier.  But that failed badly where lots of various cloud providers actually mangling the distribution images enough so it failed and required manual intervention.  Running the `openvpn3-admin init-config` as a "post installation scriptlet" solved that on essentially all the popular distributions cloud providers typically provide.

You could run `openvpn3-admin init-config` as part of the compilation phase - like after the main build has completd, if it's expected that NixOS has some kind of a "standard base setup".  The `openvpn3-admin init-config` also can also take a `--state-dir` argument pointing at a directory where to save the configs, so that can be collected into a directory where the packaging can preserve it.  Just be sure that the final installed `/var/lib/openvpn3` + `/var/lib/openvpn3/configs` directories and the files stored there have the proper file access setup.  You see the output of that command to see what it really tests for; that's what you need to ensure is correctly packaged.

Reply:
> You could run openvpn3-admin init-config as part of the compilation phase

Package builds are isolated in nixos, and as `init-config` tries to detect `systemd-resolved` presence based on dbus service existence (there is no dbus in build sandbox)/resolv.conf text (resolv.conf is replaced with empty file, as builds don't have the internet access) - it will fail to discover its existence.

You can think of it similarly to raspberry pi sdcard image builds, some machine generates .img file while not running software installed inside, and then this .img is being run on the target machine.
Except with NixOS, similar strategy is done on every system update, not only for embedded systems, and users can choose any set of packages with any configuration to install in resulting ".img".
There is no imperative package management, system config in => built system "image" out.

I think best course of action here will be generating config json files explicitly, making them read-only, and expecting that `init-config` will fail (due to configs being read-only failure) if config json files generated by us are incompatible with the new openvpn3 version, thus requiring migration on our side? (We can automate such testing in NixOS test infrastructure).
It is also possible to make user able to overlay those configs with their changes for `log_dbus_details` option, and we can make those overlays future-proof agains upstream changes, as our config overlay system allows us to rename/deprecate/remove options from overlays to make them compatible with the future config file format versions.

As the system is declarative, we can fill those configs with the correct values (We know from the system configuration if the built system will have journald/resolved) without requiring any detection in the first place.

Reply:
If you can generate two json files with some default values, that could be a workaround running the `openvpn3-admin init-config` command.

* `log-service.json`
   ```
        {
             "journald" : true,
             "log_dbus_details" : true
        }
    ```
* `netcfg.json`
    ```
        {
             "systemd_resolved" : true
        }
    ```

You might consider if it's beneficial to also boost the default log level in the log service for your users.  The default is aiming for long-term production environment and doesn't log much details (`log-level 3`).  If you add `"log_level" : "6"` to the `log-service.json`, no log events will be filtered out by the log service.  If a VPN client config contains `verb 6`, you get all the debug details instantly.

Reply:
Thanks a lot for your feedback! I've implemented this logic in the module: now the configurations are written to `/etc` at the right moment and are modifiable via module configuration.

Now marking the PR as ready for review